Document function rules for File-based Access Control (#14774)
trinodb/trino: docs/src/main/sphinx/security/file-system-access-control.rst

Configuration
-------------

.. warning::

>     Access to :doc:`table functions </functions/table>` is allowed by default.

Reviewer: Isnt it allowed for all functions?

Author: True, I wrote this warning because before access to table functions was always denied - there was no rules to control the access. Maybe change this into:
```
Access to all functions including table functions is allowed by default.
```
@mosabua WDYT?

Reviewer: yes

Configuration
-------------

> .. warning::

Reviewer 2: I think this should be part of the release notes. If we add warnings like that we may end up with messy docs.

Reviewer 3: Agreed, and it should be in a breaking change section, as this will break expected security behaviors.

Reviewer 3: I was told though that there are no special sections, and release notes are generated from issues?

Reviewer: Release notes are not generated. We write them .. but we do not have a dedicated breaking changes section. But this can be added .. probably was already .. @colebow ?

Also the warning is still appropriate .. maybe without the version info and past behavior though

Author: I remove information about Trino versions

Reviewer 2: I think we still need this too in release notes.

>     As of Trino 401 and later, access to PTFs (Polymorphic Table Function)

Reviewer 3: Functions)

Author: Changed

>     From Trino 401 the default access to PTFs (Polymorphic Table Function) has

Reviewer 4: ```suggestion
    As of Trino 401, the default access to PTFs (Polymorphic Table Function) has
```

Author: I mix this with https://github.com/trinodb/trino/pull/14774#discussion_r1009694755

>     From Trino 401 the default access to PTFs (Polymorphic Table Function) has

Author: I mix this with https://github.com/trinodb/trino/pull/14774#discussion_r1009685115

>     changed from always denied to allowed. To mitigate that it's require to add
>     a ``function`` :ref:`rule <system-file-function-rules>` to deny ``TABLE``
>     function kind.

Reviewer 4: ```suggestion
    changed from always denied to allowed. To mitigate this, you should add
    a ``function`` :ref:`rule <system-file-function-rules>` to deny ``TABLE``
    functions.
```

Reviewer 4: @m57lyra may have a better idea for how to clearly word this warning.

>     :ref:`rule <system-file-function-rules>` to deny ``TABLE`` function type.

Reviewer 3: deny the ``

Author: Changed

UPDATE                               all                update
==================================== ========== ======= ==================== ===================================================

> Permissions required for executing functions and PTFs (Polymorphic Table Function):

Reviewer 4: If you're adding a new table, please make it a list-table. You can search in the docs for `list-table` for examples and formatting.

Author: I used new `list-table`, please confirm if I did it right

> Permissions required for executing functions and PTFs (Polymorphic Table Function):

Reviewer 4: ```suggestion
Permissions required for executing functions:
```
I don't think we need to be specifically calling out PTFs - they're also functions, and while they're neat, they don't need super special treatment relative to other functions.

Author: Changed

==================================== ========== ======================== ========================== ===============================================================================
SQL command                          Catalog    Function permission      Function kind              Note
==================================== ========== ======================== ========================== ===============================================================================
SELECT function()                               execute, grant_execute*  aggregate, scalar, window  grant_execute is required when function is executed with view owner privileges

> SELECT FROM TABLE(table_function())  all        execute, grant_execute*  table                      grant_execute is required when PTF is executed with view owner privileges

Author: Changed

UPDATE                               all                update
==================================== ========== ======= ==================== ===================================================

Permissions required for executing functions:

.. list-table::
   :widths: 30, 10, 15, 15, 30
   :header-rows: 1

   * - SQL command
     - Catalog
     - Function permission
     - Function kind
     - Note
   * - ``SELECT function()``
     -
     - ``execute``, ``grant_execute*``
     - ``aggregate``, ``scalar``, ``window``
     - ``grant_execute`` is required when function is executed with view owner privileges.
   * - ``SELECT FROM TABLE(table_function())``
     - ``all``
     - ``execute``, ``grant_execute*``
     - ``table``

>      - ``grant_execute`` is required when PTF is executed with view owner privileges.

Reviewer 2: Is it well know what PTF means? I would prefer to use table function. Same in other places.

Reviewer 3: are all table functions polymorphic? If not, we need to differentiate the two. Google style guide and numerous other documentation standards specify that you spell term out the first time it is used, with the acronym. Thereafter, you just use the acronym. When things have entered very common usage (like HTTP, unless it is actual documentation for that protocol), you don't have to spell it out.

Reviewer: No .. PTF is not well known .. use "table functions" .. and @m57lyra .. not all table functions are polymorphic .. check out the video I created and shared recently

Author: Maybe make a link to https://trino.io/docs/current/functions/table.html in every occurrence of "PTF" ?

Author: I linked every `table function` to https://trino.io/docs/current/functions/table.html

visible. Specifically:

* ``catalog``: Visible if user is the owner of any nested schema, has

>   permissions on any nested table or PTF (Polymorphic Table Function) or has

Reviewer 4: ```suggestion
  permissions on any nested table or PTF (Polymorphic Table Function), or has
```

>   permissions to set session properties in the catalog.

Author: Mixed with https://github.com/trinodb/trino/pull/14774#discussion_r1009688718

* ``schema``: Visible if the user is the owner of the schema, or has permissions

>   on any nested table or PTF (Polymorphic Table Function).

Author: Changed

visible. Specifically:

* ``catalog``: Visible if user is the owner of any nested schema, has
  permissions on any nested table or PTF, or has permissions to set session properties in the catalog.
* ``schema``: Visible if the user is the owner of the schema, or has permissions

>   on any nested table or PTF.

Reviewer: table function

Author: Changed

      ]
    }

.. _system-file-function-rules:

Function rules
^^^^^^^^^^^^^^

> These rules control the ability of a user to execute SQL functions, including

Author: Mixed with https://github.com/trinodb/trino/pull/14774#discussion_r1009690830

> These rules control the ability of a user to execute SQL functions, including
> PTFs (Polymorphic Table Functions).

Reviewer 4: ```suggestion
These rules control the user's ability to execute SQL functions.
```
Same thing as above where I don't think we need to say functions _and_ PTFs.

Author: Mixed with https://github.com/trinodb/trino/pull/14774#discussion_r1009696805

Each function rule is composed of the following fields:

* ``user`` (optional): regex to match against user name. Defaults to ``.*``.
* ``role`` (optional): regex to match against role names. Defaults to ``.*``.
* ``group`` (optional): regex to match against group names. Defaults to ``.*``.
* ``catalog`` (optional): regex to match against catalog name. Defaults to
  ``.*``.
* ``schema`` (optional): regex to match against schema name. Defaults to ``.*``.
* ``function`` (optional): regex to match against function names. Defaults to
  ``.*``.
* ``privileges`` (required): zero or more of ``EXECUTE``, ``GRANT_EXECUTE``.

> * ``function_kinds`` (required): one or more of ``AGGREGATE``, ``SCALAR``,

Author: Changed

  ``TABLES``, ``WINDOW``. When user define a rule for ``AGGREGATE``, ``SCALAR``

>   or ``WINDOW`` function, ``catalog`` and ``schema`` fields are disallowed.

Author: Changed

> For example, to restore previous behavior (all Polymorphic Table Functions from any
> catalogs is denied) you can use the following rules:

Reviewer 4: ```suggestion
For example, to restore previous behavior (all polymorphic table functions from any
catalog are denied), you can use the following rules:
```

Author: I used the suggestion from https://github.com/trinodb/trino/pull/14774#discussion_r1009704916

> catalogs is denied) you can use the following rules:

Author: Changed

.. code-block:: json

    {
      "functions": [
        {
          "privileges": [
            "EXECUTE",
            "GRANT_EXECUTE"
          ],
          "function_kinds": [
            "SCALAR",
            "AGGREGATE",
            "WINDOW"
          ]
        },
        {
          "privileges": [],
          "function_kinds": [
            "TABLE"
          ]
        }
      ]
    }

> To allow executing ``query`` polymorphic table function from any catalogs by
> ``admin`` user.

Reviewer 4: ```suggestion
To allow use of the ``query`` polymorphic table function from any catalog by
an ``admin`` user:
```

Author: I used the suggestion from https://github.com/trinodb/trino/pull/14774#discussion_r1009706978

> ``admin`` user.

Author: Changed

> .. note::
> 
>     ``query`` polymorphic table function should be used by admins in a first
>     place.

Reviewer 4: Do we really need this note? It's a little jarring, but If so...

```suggestion
.. note::

    It is recommended to only allow admins to use the ``query`` polymorphic table
    function.
```

Author: Changed

>     place.

Author: > Do you mean that only admins should have this privilege?

In short yes. `query` PTF bypass the whole access control and should be used carefully.
To make access more strict, `admin` should create a `SECURITY DEFINER` `views` which execute specific query using `query` PTF so user can use only those defined by `admin`.

      ]
    }

.. _system-file-function-rules:

Function rules
^^^^^^^^^^^^^^

These rules control the user's ability to execute SQL functions,
such as :doc:`aggregate functions </functions/aggregate>`, scalar functions,
:doc:`table functions </functions/table>` and :doc:`window functions </functions/window>`.

> 

Reviewer: Somewhere here we should add that by default all functions are allowed.

Author: ```
These rules control the user's ability to execute SQL all function kinds, such as...
```
?

Reviewer: Sure .. but also say .. by default all functions are available.

Author: When user does not define any `functions` rule then all functions are available. That's how the rest of rules works as well.

  Defaults to ``.*``.
* ``schema`` (optional): regular expression to match against schema name.
  Defaults to ``.*``.
* ``function`` (optional): regular expression to match against function names.
  Defaults to ``.*``.
* ``privileges`` (required): zero or more of ``EXECUTE``, ``GRANT_EXECUTE``.
* ``function_kinds`` (required): one or more of ``AGGREGATE``, ``SCALAR``,
  ``TABLES``, ``WINDOW``. When a user defines a rule for ``AGGREGATE``, ``SCALAR``
  or ``WINDOW`` functions, the ``catalog`` and ``schema`` fields are disallowed.

To deny all :doc:`table functions </functions/table>` from any catalog,
use the following rules:

.. code-block:: json

    {
      "functions": [
        {
          "privileges": [
            "EXECUTE",
            "GRANT_EXECUTE"
          ],
          "function_kinds": [
            "SCALAR",
            "AGGREGATE",
            "WINDOW"
          ]
        },
        {
          "privileges": [],
          "function_kinds": [
            "TABLE"
          ]
        }
      ]
    }

The following example allows the ``admin`` user to execute ``query`` table
function from any catalog:

.. note::

>     It is recommended to only allow admins to use the ``query``

Reviewer: Do not make this recommendation.. change to normal paragraph that says the following section disallows usage of the query table function. And maybe add a reason why that might make sense (query passthrough ignores access controls)

Author: Got it, rewritten.

      ]
    }

.. _system-file-function-rules:

Function rules
^^^^^^^^^^^^^^

> These rules control the user's ability to execute SQL functions including PTFs.

Reviewer 3: functions, including

Author: Changed

> These rules control the user's ability to execute SQL functions, including PTFs.

Author: I added links to specific function sections.

Each function rule is composed of the following fields:

> * ``user`` (optional): ``regex`` to match against user name. Defaults to ``.*``.

Reviewer: what kind of regex? A SQL regex or a Java regex? Or something else. Also .. regex is not a keyword or so .. just say regular expression

* ``role`` (optional): ``regex`` to match against role names. Defaults to ``.*``.
* ``group`` (optional): ``regex`` to match against group names. Defaults to ``.*``.
* ``catalog`` (optional): ``regex`` to match against catalog name. Defaults to
  ``.*``.
* ``schema`` (optional): ``regex`` to match against schema name. Defaults to ``.*``.
* ``function`` (optional): ``regex`` to match against function names. Defaults to
  ``.*``.
* ``privileges`` (required): zero or more of ``EXECUTE``, ``GRANT_EXECUTE``.
* ``function_kinds`` (required): one or more of ``AGGREGATE``, ``SCALAR``,
  ``TABLES``, ``WINDOW``. When a user defines a rule for ``AGGREGATE``, ``SCALAR``
  or ``WINDOW`` functions, the ``catalog`` and ``schema`` fields are disallowed.

> To restore previous behavior, which denied all PTFs from any catalog, use the following

Reviewer 2: What is "previous"? I would prefer to write to achieve X behavior, do y configuration, where you need to explain what X is.

Reviewer: Reword this to say what the setup does.. and ignore the fact that this was the default in older releases.

Author: Reworded

rules:

.. code-block:: json

    {
      "functions": [
        {
          "privileges": [
            "EXECUTE",
            "GRANT_EXECUTE"
          ],
          "function_kinds": [
            "SCALAR",
            "AGGREGATE",
            "WINDOW"
          ]
        },
        {
          "privileges": [],
          "function_kinds": [
            "TABLE"
          ]
        }
      ]
    }

The following example allows the ``admin`` user to execute ``query`` PTFs from any
catalog:

.. note::

    It is recommended to only allow admins to use the ``query`` polymorphic table

>     function.

Reviewer 2: Add "why"

Maybe it would be nice to have some best-practices somewhere. Where you would have config to grant access to PTF to admin, then create a (definer) view on top of it with admin as owner and granting access to view to regular users. Maybe we could explain that only somewhere.

Reviewer: I would even say .. remove that.. there is no reason for this to be best practice in general. Instead explain what the consequence is of this setup .. and earlier what happens by default

Reviewer: As in like .. if anyone is allowed to execute table functions.. then what?

Author: `query` table function which works as a query pass-though - it's bypass the whole access control so IMO it's quite important to at least mention about it

Reviewer: Yeah .. mention.. but not recommend as best practice.. its just important to understand that it is kind of a security hole otherwise ;-) .. so we need to explain that and not just have a recommendation without detailing the reasoning

* ``group`` (optional): regular expression to match against group names.
  Defaults to ``.*``.
* ``catalog`` (optional): regular expression to match against catalog name.
  Defaults to ``.*``.
* ``schema`` (optional): regular expression to match against schema name.
  Defaults to ``.*``.
* ``function`` (optional): regular expression to match against function names.
  Defaults to ``.*``.
* ``privileges`` (required): zero or more of ``EXECUTE``, ``GRANT_EXECUTE``.
* ``function_kinds`` (required): one or more of ``AGGREGATE``, ``SCALAR``,
  ``TABLES``, ``WINDOW``. When a user defines a rule for ``AGGREGATE``, ``SCALAR``
  or ``WINDOW`` functions, the ``catalog`` and ``schema`` fields are disallowed.

To deny all :doc:`table functions </functions/table>` from any catalog,
use the following rules:

.. code-block:: json

    {
      "functions": [
        {
          "privileges": [
            "EXECUTE",
            "GRANT_EXECUTE"
          ],
          "function_kinds": [
            "SCALAR",
            "AGGREGATE",
            "WINDOW"
          ]
        },
        {
          "privileges": [],
          "function_kinds": [
            "TABLE"
          ]
        }
      ]
    }

> It's a good practice to limit access to ``query`` table function

Reviewer: Fix wrapping of this paragraph .. but otherwise this is good now imho

Author: Fixed, I think

    ``mask`` can contain conditional expressions such as ``IF`` or ``CASE``, which achieves conditional masking.

Function rules
^^^^^^^^^^^^^^

Each function rule is composed of the following fields:

* ``user`` (optional): regex to match against user name. Defaults to ``.*``.
* ``role`` (optional): regex to match against role names. Defaults to ``.*``.
* ``group`` (optional): regex to match against group names. Defaults to ``.*``.
* ``schema`` (optional): regex to match against schema name. Defaults to ``.*``.
* ``function`` (optional): regex to match against function names. Defaults to
  ``.*``.
* ``privileges`` (required): zero or more of ``EXECUTE``, ``GRANT_EXECUTE``.
* ``function_kinds`` (required): one or more of ``AGGREGATE``, ``SCALAR``,
  ``TABLES``, ``WINDOW``. When user define a rule for ``AGGREGATE``, ``SCALAR``

>   or ``WINDOW`` function, ``catalog`` and ``schema`` fields are disallowed.

Reviewer 3: This appears to duplicate lines 358 to 371. Is this intentional @huberty89 ? If so, all suggestions and comments there apply here also.

Author: Everything here is the same expect `role` and `catalog` - I forgot to remove `role` entry before.

    ``mask`` can contain conditional expressions such as ``IF`` or ``CASE``, which achieves conditional masking.

Function rules
^^^^^^^^^^^^^^

Each function rule is composed of the following fields:

* ``user`` (optional): regular expression to match against user name.
  Defaults to ``.*``.
* ``group`` (optional): regular expression to match against group names.
  Defaults to ``.*``.
* ``schema`` (optional): regular expression to match against schema name.
  Defaults to ``.*``.
* ``function`` (optional): regular expression to match against function names.
  Defaults to ``.*``.
* ``privileges`` (required): zero or more of ``EXECUTE``, ``GRANT_EXECUTE``.
* ``function_kinds`` (required): one or more of ``AGGREGATE``, ``SCALAR``,
  ``TABLES``, ``WINDOW``. When a user defines a rule for ``AGGREGATE``, ``SCALAR``

>   or ``WINDOW`` functions, the ``catalog`` and ``schema`` fields are disallowed.

Reviewer: should we add why ... 

because these are globally defined functions 

or something like that

Author: Ok, added explanation why
